src/Tableau.cpp

```diff
@@ -0,0 +1,395 @@
+                    tableau[i][a + nqubits] ^= tableau[i][a];
+                }
+                for (auto i = 0U; i < nqubits; i++) {
+                    tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                    std::swap(tableau[i][a], tableau[i][a + nqubits]);
+                }
+                return 4U;
+            } else {
+                const auto a = (*gate->getControls().begin()).qubit;
+                const auto b = gate->getTargets().at(0);
+                if (a == b) {
+                    util::fatal("Invalid CNOT with same control and target.");
+                }
+                for (auto i = 0U; i < nqubits; i++) {
+                    const auto xa = tableau[i][a];
+                    const auto za = tableau[i][a + nqubits];
+                    const auto xb = tableau[i][b];
+                    const auto zb = tableau[i][b + nqubits];
+                    tableau[i][2 * nqubits] ^= (xa & zb) & ((xb ^ za) ^ 1);
+                    tableau[i][a + nqubits] = za ^ zb;
+                    tableau[i][b]           = xb ^ xa;
+                }
+                return 1U;
+            }
+        }
+
+        case qc::OpType::Sdag: { // Sdag  = S x S x S
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
+            const auto a = gate->getTargets().at(0U);
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            return 3U;
+        }
+        case qc::OpType::Z: { // Z = S x S
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
+            const auto a = gate->getTargets().at(0U);
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= tableau[i][a] & tableau[i][a + nqubits];
+                tableau[i][a + nqubits] ^= tableau[i][a];
+            }
+            return 2U;
+        }
+        case qc::OpType::Y: { // Y = H x S x S x H x S x S
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
```

Comment:
as above, supporting controlled-Y is easy since it's a CNOT sandwiched by an Sdag and a S on the target.